### ISSUE TITLE
handle negative hashes 

### ISSUE THREAD

**Issue by herrcore** [OALabs/hashdb-ida#3]
Currently we use `get_highlight` to pull the selected constant from IDA but this is an issue when the constant is negative. The negative sign is not highlighted in IDA so we don't grab it and misinterpret the constant as positive. 

<img width="286" alt="Screen Shot 2021-09-26 at 9 40 22 PM" src="https://user-images.githubusercontent.com/5906222/134833526-122a2077-1244-4d5b-9873-5c79449f7de2.png">

This can be handled elegantly in the disassembly by checking the operands at the selected address for both negative and positive values of the selected constant to determine which one to use... however this doesn't really work when the value is selected in the pseudocode window.

**TLDR Issue**
When we use `get_highlight` in the pseudocode window to select a constant how can we match that constant with the actual value in the IDA microcode/pseudocode representation? This is my lack of understanding of how to manipulate the pseudocode from python... example code would be much appreciated 🙏

**Comment by anthonyprintup** [comment]
So it seems like there is a much simpler way of figuring out if the value is negative (notice the `invsign` and `negated` keywords):
![image](https://user-images.githubusercontent.com/92564080/139930646-98002656-d88a-4dc4-9814-96e0c0833a49.png)
![image](https://user-images.githubusercontent.com/92564080/139930726-5cfb7309-f0fd-4e19-b80c-64180d89de66.png)

I know that some Python code is executed when displaying the popup, which probably means IDA also gets the information through a Python API? If so, then we just need to figure out what IDA does here.

**Comment by I-VANN** [comment]
add this function and replace all "parse_highlighted_value" with "extract_immediate_value":

```
def extract_immediate_value():
        # get instruction at mouse pos
        current_ea = idaapi.get_screen_ea()

        #get instruction operand
        insn = idaapi.insn_t()
        idaapi.decode_insn(insn, current_ea)
        for op in insn.ops:
            # check if immediate value
            if op.type == idaapi.o_imm:
                # get operand value
                imm_value = op.value
                #print("TEST: Operand is an immediate: {:#x}".format(imm_value))
                return imm_value
        #print("TEST: Operand is not an immediate")
        return None
```

**Comment by anthonyprintup** [comment]
> add this function and replace all "parse_highlighted_value" with "extract_immediate_value":
> 
> ```
> def extract_immediate_value():
>         # get instruction at mouse pos
>         current_ea = idaapi.get_screen_ea()
> 
>         #get instruction operand
>         insn = idaapi.insn_t()
>         idaapi.decode_insn(insn, current_ea)
>         for op in insn.ops:
>             # check if immediate value
>             if op.type == idaapi.o_imm:
>                 # get operand value
>                 imm_value = op.value
>                 #print("TEST: Operand is an immediate: {:#x}".format(imm_value))
>                 return imm_value
>         #print("TEST: Operand is not an immediate")
>         return None
> ```

The reason why we can't use this method is that we'd have to parse the microcode instead. The highlighted values could be optimized from multiple immediate values, they could also be registers, etc.

**Comment by I-VANN** [comment]
> The reason why we can't use this method is that we'd have to parse the microcode instead. The highlighted values could be optimized from multiple immediate values, they could also be registers, etc.

it must be considered that the value of the hash is always a number and is taken from the disassembly window and not from the decompilated window where, on the other hand, there may be more immediate values. Could you please provide me with a particular use case that is not covered? Thanks.
